Add uc 08 derive simulation production configuration parameters

tobiaskleiner opened the pull request:
This PR adds the functionality defined in UC8 with the following four applications:

- simtools-production-calculate-resource-estimates
   -- calculates compute and storage resources --

- simtools-production-generate-grid
-- generates a grid of simulation points --

- simtools-production-generate-simulation-config
-- generates simulation parameters for a specific grid point --

- simtools-production-scale-events
-- metric evaluation and statistical error calculations --

Modules are stored in `production_configuration`:

- `calculate_statistical_errors_grid_point.py`
- `derive_computing_resources.py`
- `generate_production_grid.py`
- `generate_simulation_config.py`
- `interpolation_handler.py`

GernotMaier commented:
Would like start the discussions how to best organize the code. At this point most of it is in the applications, but we want to move it into modules. 

Suggest to start a module to put those classes in. What would be a good name? 

- `simtools.production_tools` (don't like `tools`, seems to cover everything)
- `simtools.production_configuration`
- `simtools.simulation_production`
- ...something shorter...?

This module would include all classes currently defined in 

- `simtools/applications/derive_resources.py` (which will be renamed to `simtools/applications/derive_computing_resources.py`
- `simtools/applications/generate_grid.py` (which will be renamed to `simtools/applications/generate_production_grid.py`)
- `simtools/applications/generate_simulation_config.py`
- `simtools/utils/calculate_statistical_errors_grid_point.py (maybe we can find a shorter name)

@orelgueta , @tobiaskleiner : please comment.

orelgueta commented:
Not much for me to comment. This is the first time I see that those modules are in applications and I definitely agree they should go into simtools instead. In terms of which name to put them under, I would vote for `simtools.production_configuration`. 
However, I would also check if we can integrate these classes with the current modules we have (at least partly).

tobiaskleiner commented:
I agree with the `simtools.production_configuration` suggestion and to move some parts into visualization or in other parts.

GernotMaier commented on simtools/production_configuration/calculate_statistical_errors_grid_point.py:
```diff
@@ -0,0 +1,719 @@
+
+        # Determine the effective area threshold (50% of max effective area)
+        max_efficiency = np.max(efficiencies)
+        threshold_efficiency = 0.1 * max_efficiency
```
The docstring says "exceeds 50%". Do you use actually 10%?

GernotMaier commented:
@tobiaskleiner -thanks a lot!

Couple of 'technical' comments from browsing once through the code:

- in general: we want to have a simple application calling the different classes (more or less what you put into the docstring) for some given input files. No 'main' in the modules (e.g., as in SimulationConfig)
- try to avoid too many hard-wired items (e.g., paths, file names)
- need to discuss if we use units in general or not (it is generally clearer with units)

calculate_statistical_errors_grid_point

- if it possible that 'effective area' is actually trigger efficiency without the multiplication by the scatter area? I at least can't find the scatter area anywhere.
- suggest to separate the file into two with one class per file (I think we always have one class per file only)

generate_grid:

- lat/long of the observatory should be read from the model parameters for each site (reference_point_latitude and reference_point_longitude)
- suggest to discuss how we treat the data levels and their corresponding accuracies; not sure if this should be hardwired or be part of a configuration input (same for high/low precision)

resource_estimates.yaml: 

- we try not to have any data files in the source code. During operation, we will retrieve this information probably from a database. I suggest for now to move it into "tests/resources" (as we are developing code and use this mostly for testing)
- suggest to do minor changes to the format and add a name to the zenith fields (dictionary). If possibly and not too much work, explicit units would be nice (and not encoded in the names)

GernotMaier commented:
@tobiaskleiner - before I start the review, could you make sure that the integration tests run successfully?

GernotMaier commented:
Running the example production_generate_simulation_config gives:

```
 python simtools/applications/production_generate_simulation_config.py  --azimuth 60.0 --elevation 45.0 --nsb 0.3 --data_level "A" --science_case "high_precision"  --file_path tests/resources/production_dl2_fits/prod6_LaPalma-20deg_gamma_cone.N.Am-4LSTs09MSTs_ID0_reduced.fits --file_type "On-source" --metrics_file  tests/resources/production_simulation_config_metrics.yaml --site North
Effective Area Error (avg): 0.000, Reference: 0.020
Signal Efficiency Error: 0.020, Reference: 0.020
INFO::calculate_statistical_errors_grid_point(l278)::calculate_error_energy_estimate_bdt_reg_tree::Calculating Energy Resolution Error
Energy Estimate Error: 0.184, Reference: 0.050
Gamma-Ray PSF Error: 0.010, Reference: 0.010
Image Template Methods Error: 0.050, Reference: 0.030
error_eff_area {'relative_errors': array([0.00000000e+00, 0.00000000e+00, 0.00000000e+00, 1.01799919e-08,
       2.21965306e-08, 3.88179482e-08, 5.76378038e-08, 7.53500614e-08,
       8.98535771e-08, 1.20081083e-07, 1.46732027e-07, 1.66036331e-07,
       2.16994290e-07, 2.42594940e-07, 2.92140457e-07, 3.35214284e-07,
       3.58574316e-07, 4.90549293e-07, 4.77865852e-07, 6.57241537e-07,
       6.62752408e-07, 6.89972899e-07, 1.03031658e-06, 1.26143022e-06,
       1.53338834e-06, 1.47603219e-06, 1.79685160e-06, 2.19451145e-06,
       1.94774912e-06, 3.34303448e-06, 4.51782360e-06, 3.89789862e-06,
       0.00000000e+00, 0.00000000e+00, 0.00000000e+00, 0.00000000e+00,
       0.00000000e+00, 0.00000000e+00, 0.00000000e+00, 0.00000000e+00,
       0.00000000e+00, 0.00000000e+00, 0.00000000e+00, 0.00000000e+00,
       0.00000000e+00, 0.00000000e+00, 0.00000000e+00, 0.00000000e+00,
       0.00000000e+00, 0.00000000e+00, 0.00000000e+00, 0.00000000e+00,
       0.00000000e+00, 0.00000000e+00, 0.00000000e+00, 0.00000000e+00,
       0.00000000e+00, 0.00000000e+00, 0.00000000e+00, 0.00000000e+00,
       0.00000000e+00, 0.00000000e+00])}
INFO::production_generate_simulation_config(l180)::main::Simulation parameters: {'core_scatter_area': np.float64(200070.0), 'viewcone': np.float64(1053.0), 'number_of_events': 11000004994}
INFO::production_generate_simulation_config(l183)::main::Simulation parameters saved to: /workdir/external/simtools/simtools-output/production_generate_simulation_config/configured_simulation_params.json
```

Is a viewcone of 1053 reasonable? I assume it is in degrees.

Here and throughput the added code: there are no units anywhere. I assume that this output is in the units expected by CORSIKA, but I am not entirely sure. The code also relies on certain units of the values read from the DL2 file, and I think at least there the units are given in the header or table columns. I suggest to use units (as we do not know if DL2 files change in future).

GernotMaier commented:
Also surprised how small the error on the effective areas area (same for the values in tests/unit_tests/production_configuration/test_generate_simulation_config.py)

GernotMaier commented:
First: great work @tobiaskleiner! This adds an almost complete framework for the configuration.

This PR is unusual big with several independent applications added, all including important . Too late to split it up, but we need to do it in pieces (I will sent a review for the first application out soon).

I have looked until now at `production_generate_simulation_config` only, which determines the configuration parameters for a single grid point.

Quite a few of the statistics questions are open, including the details of the metrics. I would suggest that you add a small discussion note on this topic into the [implementation gitlab](https://gitlab.cta-observatory.org/cta-computing/dpps/simulationpipeline/roadmap-and-implementation/-/tree/main/use-cases/use-case-08-production-configuration?ref_type=heads) (UC8) directory addressing the following question:

- what metrics do we want to consider and how are they calculated
- how to we determine from the metrics the number of required events
- how to you plan to calculate the scatter area and view cone

I think it is easier to discuss the methodical approach there and not as part of the code review. Maybe for future we should discuss the methods before the implementation starts.

Additional, we should discuss the concepts / impact of data level and science cases in with implementation gitlab. I think the science cases are not documented anywhere? What are the assumptions / motivation?

GernotMaier commented on simtools/production_configuration/calculate_statistical_errors_grid_point.py:
```diff
@@ -0,0 +1,472 @@
+
+                sim_events_data = hdul["SIMULATED EVENTS"].data  # pylint: disable=E1101
+                bin_edges_low = sim_events_data["MC_ENERG_LO"]
+                bin_edges_high = sim_events_data["MC_ENERG_HI"]
```
Is the assumption that we want to pass the metric over the energy range given by the DL2 file good? Would it make sense to define a `valid energy range` for a given metric? e.g. generate a simulation production configuration with a 0.1% statistical uncertainty in the 30 GeV to 300 TeV range (although we simulated from 10 GeV to 500 TeV?)

tobiaskleiner replied:
Already added a validity range with units to each metric in the yaml files.

tobiaskleiner commented on simtools/applications/production_generate_simulation_config.py:
```diff
@@ -0,0 +1,187 @@
+            "error_gamma_ray_psf": 0.01,
+            "error_image_template_methods": 0.03,}
+    """
+    if file_path and os.path.exists(file_path):
```
changed

tobiaskleiner commented on simtools/applications/production_generate_simulation_config.py:
```diff
@@ -0,0 +1,187 @@
+
+        Example:
+
+        metrics = {
```
changed

tobiaskleiner commented on simtools/applications/production_generate_simulation_config.py:
```diff
@@ -0,0 +1,187 @@
+        else:
+            serializable_config[key] = value
+
+    logger.info(f"Simulation parameters: {serializable_config}")
```
yes, changed

tobiaskleiner commented on simtools/applications/production_generate_simulation_config.py:
```diff
@@ -0,0 +1,187 @@
+#!/usr/bin/python3
+
+r"""
+Configure and run a simulation based on command-line arguments.
```
changed

tobiaskleiner commented on simtools/production_configuration/generate_simulation_config.py:
```diff
@@ -0,0 +1,232 @@
+generate simulation parameters for a specific grid point in a statistical error
+evaluation setup. The class considers various parameters, such as azimuth,
+elevation, and night sky background, to compute core scatter areas, viewcones,
+and the required number of simulated events.
```
changed

tobiaskleiner commented on simtools/production_configuration/generate_simulation_config.py:
```diff
@@ -0,0 +1,232 @@
+- `SimulationConfig`: Main class to handle simulation configuration for a grid point.
+  - Attributes:
+    - `grid_point` (dict): Contains azimuth, elevation, and night sky background.
+    - `data_level` (str): The data level for the simulation (e.g., 'A', 'B', 'C').
```
changed

tobiaskleiner commented on simtools/production_configuration/generate_simulation_config.py:
```diff
@@ -0,0 +1,232 @@
+        The data level (e.g., 'A', 'B', 'C') for the simulation configuration.
+    science_case : str
+        The science case for the simulation configuration.
+    file_path : str
```
changed

tobiaskleiner commented on simtools/applications/production_generate_simulation_config.py:
```diff
@@ -0,0 +1,187 @@
+    The data level for the simulation (e.g., 'A', 'B', 'C').
+science_case (str, required)
+    The science case for the simulation.
+file_path (str, required)
```
changed

tobiaskleiner commented on simtools/production_configuration/generate_simulation_config.py:
```diff
@@ -0,0 +1,232 @@
+Used to configure and
+generate simulation parameters for a specific grid point in a statistical error
+evaluation setup. The class considers various parameters, such as azimuth,
+elevation, and night sky background, to compute core scatter areas, viewcones,
```
changed

tobiaskleiner commented on simtools/production_configuration/calculate_statistical_errors_grid_point.py:
```diff
@@ -0,0 +1,472 @@
+        Type of the file, either 'On-source' or 'Offset'.
+    metrics : dict, optional
+        Dictionary of metrics to evaluate. Default is None.
+    grid_point : tuple, optional
```
Using now counts and energies throughout the class. We could discuss also if we move to strictly using units in command line parser. So far we don't require quantity format and specify the desired units in docstring as i.e. in `simulate_prod` `validate_camera_efficiency`.

tobiaskleiner commented on simtools/production_configuration/calculate_statistical_errors_grid_point.py:
```diff
@@ -0,0 +1,472 @@
+
+    def load_data_from_file(self):
+        """
+        Load data from the FITS file and return relevant arrays.
```
event lists, bin edges, and simulated events histograms. I call it simply dicts with data now.

tobiaskleiner commented on simtools/production_configuration/calculate_statistical_errors_grid_point.py:
```diff
@@ -0,0 +1,472 @@
+        self.scaled_events_gridpoint = None
+        self.energy_threshold = None
+
+    def load_data_from_file(self):
```
Ok opened an issue. See #1227

tobiaskleiner commented on simtools/production_configuration/calculate_statistical_errors_grid_point.py:
```diff
@@ -0,0 +1,472 @@
+                if self.grid_point is None:
+                    unique_azimuths = np.unique(events_data["PNT_AZ"])
+                    unique_zeniths = np.unique(events_data["PNT_ALT"])
+                    if len(unique_azimuths) == 1 and len(unique_zeniths) == 1:
```
Could do that as well, but wanted to be sure they are really unique. We might have maybe combined dl2 in the future with non flat axis other than energy?

tobiaskleiner commented on simtools/production_configuration/calculate_statistical_errors_grid_point.py:
```diff
@@ -0,0 +1,472 @@
+                    "viewcone": viewcone,
+                    "core_range": core_range,
+                }
+                if self.grid_point is None:
```
Have added some log entries in case we overwrite existing grid point values.

tobiaskleiner commented on simtools/production_configuration/generate_simulation_config.py:
```diff
@@ -0,0 +1,232 @@
+        elevation = self.grid_point.get("elevation", 0)
+        night_sky_background = self.grid_point.get("night_sky_background", 0)
+
+        # implement reading of factor from LUT
```
Already opened an issue. Trying to avoid TODOs in the code as Sonarqube complains about them.

tobiaskleiner commented on simtools/production_configuration/calculate_statistical_errors_grid_point.py:
```diff
@@ -0,0 +1,472 @@
+                        # grid point(energy, azimuth, zenith, NSB, offset)
+
+        except (FileNotFoundError, KeyError) as e:
+            logging.error(f"Error loading file {self.file_path}: {e}")
```
Added

tobiaskleiner commented on simtools/production_configuration/generate_simulation_config.py:
```diff
@@ -0,0 +1,232 @@
+        elevation = self.grid_point.get("elevation", 0)
+        night_sky_background = self.grid_point.get("night_sky_background", 0)
+
+        # implement reading of factor from LUT
```
Same as above

tobiaskleiner commented on simtools/production_configuration/generate_simulation_config.py:
```diff
@@ -0,0 +1,232 @@
+        int
+            The number of simulated events required.
+        """
+        # Obtain the statistical error evaluation metrics
```
removed

tobiaskleiner commented on simtools/production_configuration/calculate_statistical_errors_grid_point.py:
```diff
@@ -0,0 +1,472 @@
+            ref_value = self.metrics.get("error_eff_area")
+            if self.error_eff_area:
+                avg_error = np.mean(self.error_eff_area["relative_errors"])
+                print(f"Effective Area Error (avg): {avg_error:.3f}, Reference: {ref_value:.3f}")
```
Replaced. 
The average here is only a placeholder and we should have a discussion (i.e. in the opened issue) about how the error should be calculated. I.e. as we discussed last week to get the maximum errors for several predefined intervals.

tobiaskleiner commented on simtools/production_configuration/calculate_statistical_errors_grid_point.py:
```diff
@@ -0,0 +1,472 @@
+            The calculated uncertainty for image template methods.
+        """
+        # implement
+        return 0.05  # Placeholder value
```
Opened issues last week already: #1218 and #1219

tobiaskleiner commented on simtools/production_configuration/calculate_statistical_errors_grid_point.py:
```diff
@@ -0,0 +1,472 @@
+        try:
+            with fits.open(self.file_path) as hdul:
+                events_data = hdul["EVENTS"].data  # pylint: disable=E1101
+                event_energies = events_data["ENERGY"]
```
renamed

tobiaskleiner commented on simtools/production_configuration/calculate_statistical_errors_grid_point.py:
```diff
@@ -0,0 +1,472 @@
+            Dictionary with uncertainties for the file.
+        """
+        bin_edges = self.create_bin_edges()
+        triggered_event_histogram = self.compute_histogram(self.data["event_energies"], bin_edges)
```
Yes, this is wrong. It's corrected now.

tobiaskleiner commented on simtools/production_configuration/calculate_statistical_errors_grid_point.py:
```diff
@@ -0,0 +1,472 @@
+            logging.error(f"Error loading file {self.file_path}: {e}")
+        return data
+
+    def create_bin_edges(self):
```
added

tobiaskleiner commented on simtools/production_configuration/calculate_statistical_errors_grid_point.py:
```diff
@@ -0,0 +1,472 @@
+        valid = (
+            (simulated_event_counts > 0)
+            & (triggered_event_counts <= simulated_event_counts)
+            & (triggered_event_counts > 5)
```
This aplies to the really far edges outside of the energy range that we are interested. I can make a cut here or let it like this to perform the calculation over all bins if there are at least a few events and apply the energy cut later when calculating the metric.

tobiaskleiner commented on simtools/production_configuration/calculate_statistical_errors_grid_point.py:
```diff
@@ -0,0 +1,472 @@
+        # apply threshold of 5 for triggered event counts data
+        valid = (
+            (simulated_event_counts > 0)
+            & (triggered_event_counts <= simulated_event_counts)
```
Hopefully it is not happening. You would add a check for this? If so I can move it into the dl2 reader, and remove the `triggered_event_counts <= simulated_event_counts` here.

tobiaskleiner commented on simtools/production_configuration/calculate_statistical_errors_grid_point.py:
```diff
@@ -0,0 +1,472 @@
+
+    def compute_efficiency_and_errors(self, triggered_event_counts, simulated_event_counts):
+        """
+        Compute energy efficiency and its statistical error using the binomial distribution.
```
done

tobiaskleiner commented on simtools/production_configuration/calculate_statistical_errors_grid_point.py:
```diff
@@ -0,0 +1,472 @@
+        -------
+        efficiencies : array
+            Array of calculated efficiencies.
+        uncertainties : array
```
done

tobiaskleiner commented on simtools/production_configuration/calculate_statistical_errors_grid_point.py:
```diff
@@ -0,0 +1,472 @@
+        triggered_event_histogram, _ = np.histogram(event_energies, bins=bin_edges)
+        return triggered_event_histogram
+
+    def compute_efficiency_and_errors(self, triggered_event_counts, simulated_event_counts):
```
added

tobiaskleiner commented on simtools/production_configuration/calculate_statistical_errors_grid_point.py:
```diff
@@ -0,0 +1,472 @@
+        """Calculate all defined metrics as specified in self.metrics and store results."""
+        if "error_eff_area" in self.metrics:
+            self.error_eff_area = self.calculate_error_eff_area()
+            ref_value = self.metrics.get("error_eff_area")
```
done

tobiaskleiner commented on simtools/production_configuration/generate_simulation_config.py:
```diff
@@ -0,0 +1,232 @@
+            The number of simulated events required.
+        """
+        # Obtain the statistical error evaluation metrics
+        self.evaluator.calculate_metrics()
```
done

tobiaskleiner commented on simtools/applications/production_generate_simulation_config.py:
```diff
@@ -0,0 +1,187 @@
+        "--elevation", type=float, required=True, help="Elevation angle in degrees."
+    )
+    config.parser.add_argument(
+        "--nsb", type=float, required=True, help="Night sky background value."
```
added

tobiaskleiner commented on simtools/applications/production_generate_simulation_config.py:
```diff
@@ -0,0 +1,187 @@
+    config.parser.add_argument(
+        "--data_level", type=str, required=True, help="Data level (e.g., 'A', 'B', 'C')."
+    )
+    config.parser.add_argument(
```
It is an open question in the [UC8](https://gitlab.cta-observatory.org/cta-computing/dpps/simulationpipeline/roadmap-and-implementation/-/blob/main/use-cases/use-case-08-production-configuration/use-case-08.md?ref_type=heads): 

> adaption of grid axes definition on science case

> differences for grid axes definitions for different data levels (e.g., Level A, B, C)

We should discuss what exactly these are and how to define them.

tobiaskleiner commented on simtools/applications/production_generate_simulation_config.py:
```diff
@@ -0,0 +1,187 @@
+        "--nsb", type=float, required=True, help="Night sky background value."
+    )
+    config.parser.add_argument(
+        "--data_level", type=str, required=True, help="Data level (e.g., 'A', 'B', 'C')."
```
See comment above. Not yet defined and we probably need more than the 3.

tobiaskleiner commented on docs/source/user-guide/applications/production_generate_simulation_config.rst:
Added here #1233

tobiaskleiner commented:
@GernotMaier thanks for the review. I went through the comments and adressed most of them. Few more points need discussion/implementation see #1233, #1227, #1219. Will let you know when I have fixed the unit tests for another review.

tobiaskleiner commented:
@GernotMaier thanks again for the review of the first part of the PR. Went through your comments and fixed the unit/integration tests. I factored out the event scaling logic and added a file for helper functions in the production configuration folder. In there we could also move the dl2 reading part in a later stage. 
Let me know if the changes make sense and if so you could start reviewing the other parts of the PR.

GernotMaier commented on simtools/production_configuration/calculate_statistical_errors_grid_point.py:
```diff
@@ -0,0 +1,460 @@
+        max_error : float
+            Maximum relative error.
+        """
+        if "relative_errors" in self.metric_results["error_eff_area"]:
```
Trying to understand a case where `relative_errors` is not in metric_results by `error_eff_area` is filled. If I understand it correctly, both variables are always filled in `calculate_metrics`?

tobiaskleiner replied:
No currently this depends on the `production_simulation_config_metrics` file and what metrics are given there (i.e. which metric computation is required).

GernotMaier commented on simtools/production_configuration/calculate_statistical_errors_grid_point.py:
```diff
@@ -0,0 +1,460 @@
+            )
+        valid = (simulated_event_counts > 0 * u.ct) & (triggered_event_counts > 0 * u.ct)
+
+        uncertainties = np.zeros_like(triggered_event_counts) * u.ct**-0.5
```
Can you explain the '-0.5'?

tobiaskleiner replied:
good point, this was wrongly implemented and the errors should be dimensionless. Previously when keeping the units the rel error turns out with this dimension.

GernotMaier commented on simtools/production_configuration/calculate_statistical_errors_grid_point.py:
```diff
@@ -0,0 +1,460 @@
+
+        return efficiencies, relative_errors
+
+    def calculate_energy_threshold(self):
```
Suggest to replace the hardwired 10% by a variable (which default is 10%)

GernotMaier commented on simtools/production_configuration/calculate_statistical_errors_grid_point.py:
```diff
@@ -0,0 +1,460 @@
+        bin_edges = np.concatenate([bin_edges_low, [bin_edges_high[-1]]])
+        return np.unique(bin_edges)
+
+    def compute_histogram(self, event_energies_reco, bin_edges):
```
Suggest 'compute_triggered_event_histogram' to make the purpose of this function clearer.

tobiaskleiner replied:
done

GernotMaier commented on simtools/production_configuration/calculate_statistical_errors_grid_point.py:
```diff
@@ -0,0 +1,460 @@
+        Parameters
+        ----------
+        event_energies_reco : array
+            Array of energies of the observed events.
```
`Array of reconstructed energy per event`

tobiaskleiner replied:
done

GernotMaier reviewed (commented):
A couple of more comments on the `production_generate_simulation_config`.

I will talk to you directly about a couple of points.

tobiaskleiner commented on simtools/applications/production_generate_simulation_config.py:
```diff
@@ -0,0 +1,156 @@
+        "--science_case", type=str, required=True, help="Science case for the simulation."
+    )
+    config.parser.add_argument(
+        "--file_path", type=str, required=True, help="Path to the dl2_mc_events_file FITS file."
```
Adding some comments here, but the changes are done in the separate part1 PR.
Changed the comment to your suggestion.

tobiaskleiner commented on simtools/applications/production_generate_simulation_config.py:
```diff
@@ -0,0 +1,156 @@
+        "--metrics_file",
+        required=True,
+        type=str,
+        help="Path to YAML file containing metrics and required precision as values (required).",
```
Done

tobiaskleiner commented on simtools/production_configuration/production_configuration_helper_functions.py:
```diff
@@ -0,0 +1,27 @@
+        Dictionary of metrics loaded from the YAML file.
+    """
+    if file_path and os.path.exists(file_path):
+        return gen.collect_data_from_file_or_dict(file_name=file_path, in_dict=True)
```
Changed

tobiaskleiner commented on simtools/production_configuration/production_configuration_helper_functions.py:
Since I am using it in 8 different places, I thought a small helper function would be better readable, but I can change it back to this if you prefer.

tobiaskleiner commented on simtools/applications/production_generate_simulation_config.py:
```diff
@@ -0,0 +1,156 @@
+        "--nsb", type=float, required=True, help="Night sky background in units of 1/(sr*ns*cm**2)."
+    )
+    config.parser.add_argument(
+        "--ctao_data_level", type=str, required=True, help="Data level (e.g., 'A', 'B', 'C')."
```
ctao_data_level is what you suggested in one of your last comments.

tobiaskleiner commented on simtools/applications/production_generate_simulation_config.py:
```diff
@@ -0,0 +1,156 @@
+
+    for key, value in simulation_params.items():
+        if isinstance(value, u.Quantity):
+            serializable_config[key] = f"{value.value} {value.unit}"
```
Currently only number_of_events is implemented. See the open issues for the viewcone and core scatter area. This relates also to your following comments.

tobiaskleiner commented on simtools/production_configuration/generate_simulation_config.py:
```diff
@@ -0,0 +1,193 @@
+        """
+        return self.event_scaler.scale_events()
+
+    def _calculate_core_scatter_area(self) -> float:
```
Unit (m2) would be added with the implementation.

tobiaskleiner commented on simtools/production_configuration/generate_simulation_config.py:
```diff
@@ -0,0 +1,193 @@
+        float
+            The area factor.
+        """
+        azimuth = self.grid_point.get("azimuth", 0)
```
See comment above (only placeholder here)

tobiaskleiner commented on simtools/production_configuration/calculate_statistical_errors_grid_point.py:
```diff
@@ -0,0 +1,460 @@
+    ----------
+    file_path : str
+        Path to the dl2_mc_events_file FITS file.
+    file_type : str
```
changed to point-like or cone

tobiaskleiner commented on simtools/production_configuration/calculate_statistical_errors_grid_point.py:
```diff
@@ -0,0 +1,460 @@
+                        sim_units[col_name] = None
+
+                # Check and apply units to each column, raising an error if the unit is missing
+                if not event_units["ENERGY"]:
```
removed

tobiaskleiner commented on simtools/production_configuration/generate_simulation_config.py:
```diff
@@ -0,0 +1,193 @@
+        self.file_path = file_path
+        self.file_type = file_type
+        self.metrics = metrics or {}
+        self.evaluator = StatisticalErrorEvaluator(file_path, file_type, metrics)
```
As discussed I will remove the option to override the grid point for a file from outside of the instance. This requires that we have all the parameters in the dl2 file for setting the grid point (which is the case).

tobiaskleiner commented on simtools/production_configuration/calculate_statistical_errors_grid_point.py:
```diff
@@ -0,0 +1,460 @@
+                    if len(unique_azimuths) == 1 and len(unique_zeniths) == 1:
+                        _logger.info(
+                            f"Setting initial grid point with azimuth: {unique_azimuths[0]}"
+                            f" zenith: {unique_zeniths[0]}"
```
See comment above

tobiaskleiner commented on simtools/production_configuration/calculate_statistical_errors_grid_point.py:
```diff
@@ -0,0 +1,460 @@
+
+    def calculate_metrics(self):
+        """Calculate all defined metrics as specified in self.metrics and store results."""
+        print("self.metrics: ", self.metrics)
```
this was a relict from debugging, removed

tobiaskleiner commented:
@GernotMaier the only part that is not yet included in the code or in the other PR is `generate_production_grid`. But maybe its best if we close this one and I open another PR with just this part included?

GernotMaier commented:
Agree. Please close the PR and put a comment on where the updated work is taking place.

tobiaskleiner commented:
Closing this PR as its parts are now either implemented (threshold/limits calculation for production, event scaling for production) or moved into other PRs:#1254 #1388.
